### ISSUE TITLE
Doing a `sum` (maybe windowed) of `Bool` value type doesn't work

### ISSUE THREAD

**Issue by tpgillam** [invenia/TimeDag.jl#22]
More generally, if (in terms of types) `T + T -> S`, where `S != T`, this doesn't work.

Fix should be easy for `InceptionSum`. For `WindowSum` it may a tiny bit harder. `AssociativeWindowAggregation.jl` almost certainly assumes that the associative combination is of the form `T * T -> T`, for arbitrary operation `*`.

**Comment by iamed2** [comment]
See also:
```
help?> Base.add_sum
  Base.add_sum(x, y)


  The reduction operator used in sum. The main difference from + is that small
  integers are promoted to Int/UInt.
```

You can get the output type for an addition by performing an operation, e.g.
```
julia> let T = Bool
           typeof(oneunit(T) + oneunit(T))
       end
Int64

julia> let T = Int16
           typeof(oneunit(T) + oneunit(T))
       end
Int16

julia> let T = Bool
           typeof(Base.add_sum(oneunit(T), oneunit(T)))
       end
Int64

julia> let T = Int16
           typeof(Base.add_sum(oneunit(T), oneunit(T)))
       end
Int64
```
There are compiler-based ways to do this but they are explicitly discouraged as they can introduce a lot of recompilation or even compiler failures into your program:
```
help?> Base.promote_op
  promote_op(f, argtypes...)


  Guess what an appropriate container eltype would be for storing results of f(::argtypes...). The
  guess is in part based on type inference, so can change any time.

  │ Warning
  │
  │  Due to its fragility, use of promote_op should be avoided. It is preferable to base the
  │  container eltype on the type of the actual elements. Only in the absence of any elements
  │  (for an empty result container), it may be unavoidable to call promote_op.
```

BTW `oneunit` was an example but you can use `zero` as well (the additive identity). I have seen both. For multiplication you'd probably want to make one of the operands `one`, which is the multiplicative identity. The difference comes up in unitful quantities, e.g.:

```
julia> mm_type = typeof(1mm)
Quantity{Int64, 𝐋, Unitful.FreeUnits{(mm,), 𝐋, nothing}}

julia> one(mm_type) + one(mm_type)
2

julia> oneunit(mm_type) + oneunit(mm_type)
2 mm

julia> zero(mm_type) + zero(mm_type)
0 mm

julia> oneunit(mm_type) * oneunit(mm_type)
1 mm²

julia> oneunit(mm_type) * one(mm_type)
1 mm

julia> zero(mm_type) * zero(mm_type)
0 mm²
```

**Comment by tpgillam** [comment]
Yep this is a good point. For the way it's designed, TimeDag requires that a function with the (desired) behaviour of `Base.promote_op` exists (*).

I'm fixing the original issue (bools not working in window sums) in https://github.com/invenia/TimeDag.jl/pull/31 . But I'll create another issue regarding the use of `Base.promote_op` internally. I can't trivially swap this out to use `oneunit` in all cases, since many types don't define methods for this function (e.g. `DateTime` is the first failure I saw in the test suite when I tried)

The main failure mode that I can think of is if we create a node in the graph based around a function with signature `f(S, T)::U`, but then we later define a method `f(S, T)::V`. I'm not sure what exactly happens right now if this happens. But I think this is still a problem if one uses the `oneunit` approach. 

(*) Or, at least, is defined for all functions and node value types with which anyone wishes to use the library.
